test/Chainweb/Test/Pact/ForkTest.hs

```diff
@@ -0,0 +1,342 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Chainweb.Test.Pact.ForkTest
+  ( test
+  , scheduledTest
+  ) where
+
+import Control.Concurrent.Async
+import Control.Concurrent.MVar
+import Control.Concurrent.STM.TBQueue
+import Control.Exception (bracket, throwIO)
+import Control.Lens
+import Control.Monad.IO.Class
+import Control.Monad.STM
+
+import qualified Data.Aeson as A
+import Data.CAS.HashMap hiding (toList)
+import qualified Data.HashMap.Strict as HM
+import Data.List
+import Data.Numbers.Primes
+import Data.String.Conv (toS)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Vector (Vector)
+import qualified Data.Vector as V
+
+import GHC.Natural
+
+import System.IO.Extra
+import System.LogLevel
+
+import Test.QuickCheck hiding ((.&.))
+import Test.Tasty
+import Test.Tasty.QuickCheck
+
+import Pact.Types.ChainMeta
+import Pact.Parse
+import qualified Pact.Types.ChainId as P
+import qualified Pact.Types.Command as P
+import qualified Pact.Types.Hash as P
+import qualified Pact.Types.Exp as P
+import qualified Pact.Types.PactValue as P
+
+-- internal modules
+import Chainweb.BlockHash
+import Chainweb.BlockHeader
+import Chainweb.BlockHeader.Genesis
+import Chainweb.BlockHeaderDB hiding (withBlockHeaderDb)
+import Chainweb.Graph
+import Chainweb.Logger
+import Chainweb.Miner.Pact
+import Chainweb.Pact.Backend.Types
+import qualified Chainweb.Pact.PactService as PS
+import Chainweb.Pact.Service.BlockValidation
+import Chainweb.Pact.Service.PactQueue
+import Chainweb.Pact.Service.Types
+import Chainweb.Payload
+import Chainweb.Payload.PayloadStore.Types
+import Chainweb.Test.Pact.Utils
+import Chainweb.Test.Utils
+import Chainweb.Time
+import Chainweb.Transaction
+import qualified Chainweb.TreeDB as TDB
+import Chainweb.Version
+
+test :: TestTree
+test =
+    withRocksResource $ \rocksIO ->
+    withBlockHeaderDb rocksIO _genBlock $ \bhdb ->
+    withPayloadDb $ \pdb ->
+    testProperty "prop_forkValidates" (prop_forkValidates pdb bhdb cid _genBlock)
+  where
+    _genBlock = genesisBlockHeader testVer cid
+    cid = someChainId testVer
+
+scheduledTest :: ScheduledTest
+scheduledTest = ScheduledTest "Pact checkpointer forking test" test
+
+testVer :: ChainwebVersion
+testVer = FastTimedCPM petersonChainGraph
+
+-- | Property: Fork requiring checkpointer rewind validates properly
+prop_forkValidates
+    :: IO (PayloadDb HashMapCas)
+    -> (IO BlockHeaderDb)
+    -> ChainId
+    -> BlockHeader
+    -> Property
+prop_forkValidates pdb bhdb cid genBlock =
+    again $ ioProperty $ do
+        (trunk, left, right) <- generate genForkLengths
+        mVar <- newMVar (0 :: Int)
+        withPactProp testVer Warn pdb bhdb (testMemPoolAccess cid mVar) (return Nothing) $ \reqQ -> do
+            db <- bhdb
+            putStrLn $ "Testing fork lengths:"
```

Comment:
change `putStrLn` to `debug` so you can have

```
debug :: String -> IO ()
#define DEBUG_TEST 0
#if DEBUG_TEST
debug = putStrLn
#else
debug = const $ return ()
#endif
```

Everyone grouses about chatty tests but the debug output is useful when you're trying to diagnose an issue